backends/dpdk/dpdkContext.cpp

```diff
@@ -0,0 +1,582 @@
+        auto prop = tbl->properties->getProperty(IR::TableProperties::defaultActionPropertyName);
+        if (prop && prop->isConstant)
+            has_constant_default_action = true;
+        bool is_const_default_action = false;
+        bool can_be_hit_action = true;
+        bool is_compiler_added_action = false;
+        bool can_be_default_action = !has_constant_default_action;
+
+        // First, check for action annotations
+        auto table_only_annot = action_decl->annotations->getSingle("tableonly");
+        auto default_only_annot = action_decl->annotations->getSingle("defaultonly");
+        auto hidden = action_decl->annotations->getSingle(IR::Annotation::hiddenAnnotation);
+
+        if (table_only_annot) {
+            can_be_default_action = false;
+        }
+        if (default_only_annot) {
+            can_be_hit_action = false;
+        }
+        if (hidden) {
+            is_compiler_added_action = true;
+        }
+
+        // Second, see if this action is the default action and/or constant default action
+        auto default_action = tbl->getDefaultAction();
+        if (default_action) {
+            if (auto mc = default_action->to<IR::MethodCallExpression>()) {
+                default_action = mc->method;
+            }
+
+            auto path = default_action->to<IR::PathExpression>();
+                if (!path)
+                    BUG("Default action path %s cannot be found", default_action);
+            auto actName = refmap->getDeclaration(path->path, true)->getName();
+
+            if (actName == act->getName()) {
+                if (has_constant_default_action) {
+                    can_be_default_action = true;  // this is the constant default action
+                    is_const_default_action = true;
+                }
+            }
+        }
+
+        /* DPDK target takes a structure as parameter for Actions. So, all action
+           parameters are collected into a structure by an earlier pass. */
+        auto params = ::get(structure->args_struct_map, act->externalName() + "_arg_t");
+        if (params)
+            attr.params = params->clone();
+        else
+            attr.params = nullptr;
+
+        attr.constant_default_action = is_const_default_action;
+        attr.is_compiler_added_action = is_compiler_added_action;
+        attr.allowed_as_hit_action = can_be_hit_action;
+        attr.allowed_as_default_action = can_be_default_action;
+        attr.actionHandle = getNewActionHandle();
+        actionAttrMap.emplace(act->getName(), attr);
+    }
+}
+
+/* Print a single table object of match_table type in Context JSON*/
+void WriteContextJson::printTableCtxtJson(const IR::P4Table *tbl, std::ostream &out) {
+    bool hasActionProfileSelector = false;
+    auto mainTableAttr = ::get(tableAttrmap, tbl->name.originalName);
+    cstring tableName = mainTableAttr.controlName + "." + tbl->name.originalName;
+    add_space(out, 8); out << "{\n";
```

Comment:
this is a big ugly, it would be nicer to have indentation maintained automatically and automatically increased on blocks or arrays. But I won't be upset if you don't change this.
We already have several places where we write json. Maybe we should have a common JsonWriter class. Alternatively, you could generate a JSON object and just serialize it as a string.

Reply:
Ack. I will try to reorganize the code with some additional functions to maintain the indentation and write to json.

Reply:
Done